src/main/java/build/buildfarm/worker/OutputDirectory.java

```diff
@@ -0,0 +1,82 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.worker;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Stack;
+
+// would be great to be able to inherit from ImmutableMap...
+public class OutputDirectory extends HashMap<String, OutputDirectory> {
+  private OutputDirectory() {
+  }
+
+  public static OutputDirectory parse(Iterable<String> outputFiles, Iterable<String> outputDirs) {
+    return parseDirectories(Iterables.mergeSorted(
+        ImmutableList.<Iterable<String>>of(
+            Iterables.transform(
+                outputFiles,
+                (file) -> file.contains("/") ? file.substring(0, file.lastIndexOf('/') + 1) : ""),
+            outputDirs),
+        (a, b) -> a.compareTo(b)));
+  }
+
+  private static OutputDirectory parseDirectories(Iterable<String> outputDirs) {
+    OutputDirectory outputDirectory = new OutputDirectory();
+    Stack<OutputDirectory> stack = new Stack<>();
+
+    OutputDirectory currentOutputDirectory = outputDirectory;
+    String prefix = "";
+    for (String outputDir : outputDirs) {
+      while (!outputDir.startsWith(prefix)) {
```

Comment:
I believe you were referring to these below in your comments about wanting to have these values in the test: I'll have it generate this list with a file 'a/file', in addition to the other two directories, but since 'a' is a parent of 'a/b/c', that violates our preconditions and the action would be rejected prior to this